lfs/ntlm.go

```diff
@@ -0,0 +1,215 @@
+
+	if len(splits) != 2 {
+		errorMessage := fmt.Sprintf("Your user name must be of the form DOMAIN\\user. It is currently %s", creds["username"], "string")
+		return nil, errors.New(errorMessage)
+	}
+
+	session, err := ntlm.CreateClientSession(ntlm.Version2, ntlm.ConnectionOrientedMode)
+
+	if err != nil {
+		return nil, err
+	}
+
+	session.SetUserInfo(splits[1], creds["password"], strings.ToUpper(splits[0]))
+	c.ntlmSession = session
+	return session, nil
+}
+
+func DoNTLMRequest(request *http.Request, retry bool) (*http.Response, error) {
+	handReq, err := cloneRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := InitHandShake(handReq)
+
+	if err != nil && res == nil {
+		return nil, err
+	}
+
+	//If the status is 401 then we need to re-authenticate, otherwise it was successful
+	if res.StatusCode == 401 {
+
+		creds, err := getCredsForAPI(request)
+		if err != nil {
+			return nil, err
+		}
+
+		negotiateReq, err := cloneRequest(request)
+		if err != nil {
+			return nil, err
+		}
+
+		challengeMessage, err := negotiate(negotiateReq, getNegotiateMessage())
+		if err != nil {
+			return nil, err
+		}
+
+		challengeReq, err := cloneRequest(request)
+		if err != nil {
+			return nil, err
+		}
+
+		res, err := challenge(challengeReq, challengeMessage, creds)
+		if err != nil {
+			return nil, err
+		}
+
+		//If the status is 401 then we need to re-authenticate
+		if res.StatusCode == 401 && retry == true {
+			return DoNTLMRequest(challengeReq, false)
+		}
+
+		saveCredentials(creds, res)
+
+		return res, nil
+	}
+	return res, nil
+}
+
+func InitHandShake(request *http.Request) (*http.Response, error) {
+	return Config.HttpClient().Do(request)
+}
+
+func negotiate(request *http.Request, message string) ([]byte, error) {
+	request.Header.Add("Authorization", message)
+	res, err := Config.HttpClient().Do(request)
+
+	if res == nil && err != nil {
+		return nil, err
+	}
+
+	ret, err := parseChallengeResponse(res)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer res.Body.Close()
+	defer io.Copy(ioutil.Discard, res.Body)
+
+	return ret, nil
+}
+
+func challenge(request *http.Request, challengeBytes []byte, creds Creds) (*http.Response, error) {
+	challenge, err := ntlm.ParseChallengeMessage(challengeBytes)
+
+	if err != nil {
+		return nil, err
+	}
+
+	session, err := Config.ntlmClientSession(creds)
+	if err != nil {
+		return nil, err
+	}
+
+	session.ProcessChallengeMessage(challenge)
+	authenticate, err := session.GenerateAuthenticateMessage()
+
+	if err != nil {
+		return nil, err
+	}
+
+	authenticateMessage := concatS("NTLM ", base64.StdEncoding.EncodeToString(authenticate.Bytes()))
+	request.Header.Add("Authorization", authenticateMessage)
+	return Config.HttpClient().Do(request)
+}
+
+func parseChallengeResponse(response *http.Response) ([]byte, error) {
+	header := response.Header.Get("Www-Authenticate")
+
+	//parse out the "NTLM " at the beginning of the response
+	challenge := header[5:]
```

Comment:
This can panic if the header value is too short. I'd definitely check the string length and return an error. Much better than a scary panic backtrace.